apps/src/publicKeyCryptography/Alice.jsx

```diff
@@ -0,0 +1,106 @@
+/** @file The Alice character panel from the crypto widget */
+import React from 'react';
+import CollapsiblePanel from './CollapsiblePanel';
+import NumberedSteps from './NumberedSteps';
+import IntegerField from './IntegerField';
+import IntegerTextbox from './IntegerTextbox';
+import {PrivateKeyDropdown, PublicModulusDropdown} from './cryptographyFields';
+import {computePublicKey} from './cryptographyMath';
+
+const Alice = React.createClass({
+  propTypes: {
+    setPublicModulus: React.PropTypes.func.isRequired,
+    setPublicKey: React.PropTypes.func.isRequired
+  },
+
+  getInitialState() {
+    return {
+      publicModulus: null,
+      privateKey: null,
+      publicNumber: null,
+      secretNumber: null
+    };
+  },
+
+  setPublicModulus(publicModulus) {
+    this.setState({publicModulus});
+    this.setPrivateKey(null);
+    this.clearSecretNumber();
+  },
+
+  onPublicModulusChange(publicModulus) {
+    this.setPublicModulus(publicModulus);
+    this.props.setPublicModulus(publicModulus);
+  },
+
+  setPrivateKey(privateKey) {
+    this.setState({privateKey});
+    this.clearSecretNumber();
+  },
+
+  onPrivateKeyChange(privateKey) {
+    const {publicModulus} = this.state;
+    this.setPrivateKey(privateKey);
+    this.props.setPublicKey(this.getPublicKey({privateKey, publicModulus}));
+  },
+
+  setPublicNumber(publicNumber) {
+    this.setState({publicNumber});
+    this.clearSecretNumber();
+  },
+
+  getPublicKey({privateKey, publicModulus}) {
+    return privateKey && publicModulus ? computePublicKey(privateKey, publicModulus) : null;
+  },
+
+  computeSecretNumber() {
+    const {publicModulus, privateKey, publicNumber} = this.state;
+    const secretNumber = [publicModulus, privateKey, publicNumber].every(Number.isInteger) ?
```

Comment:
It's covered by the polyfill.  That should be provided by the code-studio bundle when this is running in the Code.org context, but I should confirm that since I haven't included the polyfill in the publicKeyCryptography bundle directly.